File: apps/web/src/stores/service.ts
import { defineStore } from 'pinia'
import { ref, watch } from 'vue'
import api from '@/lib/api'
import { useLocationStore } from './location'

export interface Service {
  id: string
  organization_id: string
  location_id: string
  name: string
  active: boolean
  price_mxn: number | null
  duration_minutes: number
  sort_order: number
  created_at: string
  updated_at: string
}

export interface ServiceListResponse {
  items: Service[]
  total: number
}

export interface ServiceCreate {
  name: string
  price_mxn?: number | null
  duration_minutes?: number
  sort_order?: number
}

export interface ServiceUpdate {
  name?: string
  price_mxn?: number | null
  duration_minutes?: number
  sort_order?: number
}

export const useServiceStore = defineStore('service', () => {
  const locationStore = useLocationStore()

  const services = ref<Service[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchServices() {
    if (!locationStore.selectedLocationId) {
      services.value = []
      return
    }

    loading.value = true
    error.value = null
    try {
      const response = await api.get<ServiceListResponse>(
        `/api/v1/locations/${locationStore.selectedLocationId}/services`,
        { params: { include_inactive: true } }
      )
      services.value = response.data.items
    } catch (err) {
      error.value = 'Error al cargar servicios'
      console.error('Failed to fetch services:', err)
    } finally {
      loading.value = false
    }
  }

  async function createService(data: ServiceCreate) {
    if (!locationStore.selectedLocationId) {
      throw new Error('No hay sucursal seleccionada')
    }

    const response = await api.post<Service>(
      `/api/v1/locations/${locationStore.selectedLocationId}/services`,
      data
    )
    services.value.push(response.data)
    return response.data
  }

  async function updateService(id: string, data: ServiceUpdate) {
    const response = await api.patch<Service>(`/api/v1/services/${id}`, data)
    const index = services.value.findIndex((s) => s.id === id)
    if (index !== -1) {
      services.value[index] = response.data
    }
    return response.data
  }

  async function activateService(id: string) {
    const response = await api.post<Service>(`/api/v1/services/${id}/activate`)
    const index = services.value.findIndex((s) => s.id === id)
    if (index !== -1) {
      services.value[index] = response.data
    }
    return response.data
  }

  async function deactivateService(id: string) {
    const response = await api.post<Service>(`/api/v1/services/${id}/deactivate`)
    const index = services.value.findIndex((s) => s.id === id)
    if (index !== -1) {
      services.value[index] = response.data
    }
    return response.data
  }

  // Refetch when location changes
  watch(
    () => locationStore.selectedLocationId,
    () => {
      fetchServices()
    },
    { immediate: true }
  )

  return {
    services,
    loading,
    error,
    fetchServices,
    createService,
    updateService,
    activateService,
    deactivateService,
  }
})

File: apps/api/src/washflow_api/schemas/service.py
"""Service schemas."""

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class ServiceCreate(BaseModel):
    """Service creation schema."""

    name: str = Field(min_length=2, max_length=60)
    price_mxn: int | None = Field(default=None, ge=0)
    duration_minutes: int = Field(default=20, ge=1, le=600)
    sort_order: int = Field(default=0, ge=0)

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str) -> str:
        return v.strip()


class ServiceUpdate(BaseModel):
    """Service update schema."""

    name: str | None = Field(default=None, min_length=2, max_length=60)
    price_mxn: int | None = Field(default=None, ge=0)
    duration_minutes: int | None = Field(default=None, ge=1, le=600)
    sort_order: int | None = Field(default=None, ge=0)

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str | None) -> str | None:
        if v is not None:
            return v.strip()
        return v


class ServiceOut(BaseModel):
    """Service response schema."""

    id: str
    organization_id: str
    location_id: str
    name: str
    active: bool
    price_mxn: int | None
    duration_minutes: int
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ServiceListResponse(BaseModel):
    """Paginated service list response."""

    items: list[ServiceOut]
    total: int
